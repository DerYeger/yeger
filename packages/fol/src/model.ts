export type VariableAssignment = Record<string, number>

export type UnsafeRecord<T extends string | number, U> = Partial<Record<T, U>>

export type Constants = UnsafeRecord<string, number>
type UnaryFunctions = UnsafeRecord<string, UnsafeRecord<number, number>>

type BinaryFunctions = UnsafeRecord<
  string,
  UnsafeRecord<number, UnsafeRecord<number, number>>
>

type UnaryRelations = UnsafeRecord<string, Set<number>>

type BinaryRelations = UnsafeRecord<string, UnsafeRecord<number, Set<number>>>

export interface Model {
  readonly nodes: Set<number>
  readonly functions: {
    readonly binary: BinaryFunctions
    readonly constants: Constants
    readonly unary: UnaryFunctions
  }
  readonly relations: {
    readonly binary: BinaryRelations
    readonly unary: UnaryRelations
  }
}

export interface TreeNode {
  symbol: string
  children: TreeNode[] | undefined
}

export interface FOLFragment {
  toFormattedString(): string
}

export interface Formula extends FOLFragment {
  evaluate(model: Model, variableAssignment: VariableAssignment): boolean
  toTree(): TreeNode
}

export class ParenthesizedFormula implements Formula {
  public constructor(public readonly inner: Formula) {}

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    return this.inner.evaluate(model, variableAssignment)
  }

  public toFormattedString(): string {
    return `(${this.inner.toFormattedString()})`
  }

  public toTree(): TreeNode {
    return {
      symbol: '()',
      children: [this.inner.toTree()],
    }
  }
}

export abstract class BinaryFormula implements Formula {
  public constructor(
    public readonly left: Formula,
    public readonly operator: string,
    public readonly right: Formula
  ) {}

  public abstract evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean

  public toFormattedString(): string {
    return `${this.left.toFormattedString()} ${
      this.operator
    } ${this.right.toFormattedString()}`
  }

  public abstract toTree(): TreeNode
}

const orSymbol = '\u2228'
export class OrFormula extends BinaryFormula {
  public constructor(left: Formula, right: Formula) {
    super(left, orSymbol, right)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    return (
      this.left.evaluate(model, variableAssignment) ||
      this.right.evaluate(model, variableAssignment)
    )
  }

  public toTree(): TreeNode {
    return {
      symbol: orSymbol,
      children: [this.left.toTree(), this.right.toTree()],
    }
  }
}

const andSymbol = '\u2227'
export class AndFormula extends BinaryFormula {
  public constructor(left: Formula, right: Formula) {
    super(left, andSymbol, right)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    return (
      this.left.evaluate(model, variableAssignment) &&
      this.right.evaluate(model, variableAssignment)
    )
  }

  public toTree(): TreeNode {
    return {
      symbol: andSymbol,
      children: [this.left.toTree(), this.right.toTree()],
    }
  }
}

const implSymbol = '\u2192'
export class ImplFormula extends BinaryFormula {
  public constructor(left: Formula, right: Formula) {
    super(left, implSymbol, right)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    return (
      !this.left.evaluate(model, variableAssignment) ||
      this.right.evaluate(model, variableAssignment)
    )
  }

  public toTree(): TreeNode {
    return {
      symbol: implSymbol,
      children: [this.left.toTree(), this.right.toTree()],
    }
  }
}

const biImplSymbol = '\u2194'
export class BiImplFormula extends BinaryFormula {
  public constructor(left: Formula, right: Formula) {
    super(left, biImplSymbol, right)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    const left = this.left.evaluate(model, variableAssignment)
    const right = this.right.evaluate(model, variableAssignment)
    return (left && right) || (!left && !right)
  }

  public toTree(): TreeNode {
    return {
      symbol: biImplSymbol,
      children: [this.left.toTree(), this.right.toTree()],
    }
  }
}

export abstract class UnaryFormula implements Formula {
  public constructor(
    public readonly inner: Formula,
    public readonly operator: string
  ) {}

  public abstract evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean

  public toFormattedString(): string {
    return `${this.operator}(${this.inner.toFormattedString()})`
  }

  public abstract toTree(): TreeNode
}

const notSymbol = '\u00AC'
export class NotFormula extends UnaryFormula {
  public constructor(inner: Formula) {
    super(inner, notSymbol)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    return !this.inner.evaluate(model, variableAssignment)
  }

  public toTree(): TreeNode {
    return {
      symbol: notSymbol,
      children: [this.inner.toTree()],
    }
  }
}

export abstract class QuantorFormula implements Formula {
  public constructor(
    public readonly variable: BoundVariable,
    public readonly inner: Formula,
    public readonly quantor: string
  ) {}

  public abstract evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean

  public toFormattedString(): string {
    return `${
      this.quantor
    } ${this.variable.toFormattedString()}. ${this.inner.toFormattedString()}`
  }

  public abstract toTree(): TreeNode
}

const universalQuantorSymbol = '\u2200'
export class UniversalQuantorFormula extends QuantorFormula {
  public constructor(variable: BoundVariable, inner: Formula) {
    super(variable, inner, universalQuantorSymbol)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    for (const node of model.nodes.values()) {
      const newAssignment = {
        ...variableAssignment,
        [this.variable.name]: node,
      }
      if (!this.inner.evaluate(model, newAssignment)) {
        return false
      }
    }
    return true
  }

  public toTree(): TreeNode {
    return {
      symbol: universalQuantorSymbol,
      children: [this.inner.toTree()],
    }
  }
}

const existentialQuantorSymbol = '\u2203'
export class ExistentialQuantorFormula extends QuantorFormula {
  public constructor(variable: BoundVariable, inner: Formula) {
    super(variable, inner, existentialQuantorSymbol)
  }

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    for (const node of model.nodes.values()) {
      const newAssignment = {
        ...variableAssignment,
        [this.variable.name]: node,
      }
      if (this.inner.evaluate(model, newAssignment)) {
        return true
      }
    }
    return false
  }

  public toTree(): TreeNode {
    return {
      symbol: existentialQuantorSymbol,
      children: [this.inner.toTree()],
    }
  }
}

export interface Relation extends Formula {}

export class BooleanLiteral implements Formula {
  private constructor(
    public readonly stringRepresentation: string,
    public readonly value: boolean
  ) {}

  public evaluate(): boolean {
    return this.value
  }

  public toFormattedString(): string {
    return this.stringRepresentation
  }

  public toTree(): TreeNode {
    return {
      symbol: this.toFormattedString(),
      children: [],
    }
  }

  public static readonly True = new BooleanLiteral('tt', true)
  public static readonly False = new BooleanLiteral('ff', false)
}

export class UnaryRelation implements Relation {
  public constructor(
    public readonly name: string,
    public readonly expression: Expression
  ) {}

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    const relation = model.relations.unary[this.name]
    if (!relation) {
      throw new Error(`Missing unary relation ${this.name}`)
    }
    const interpreted = this.expression.interpret(model, variableAssignment)
    return relation.has(interpreted)
  }

  public toFormattedString(): string {
    return `${this.name}(${this.expression.toFormattedString()})`
  }

  public toTree(): TreeNode {
    return {
      symbol: this.toFormattedString(),
      children: [],
    }
  }
}

export class BinaryRelation implements Relation {
  public constructor(
    public readonly name: string,
    public readonly firstExpression: Expression,
    public readonly secondExpression: Expression
  ) {}

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    const relation = model.relations.binary[this.name]
    if (!relation) {
      throw new Error(`Missing binary relation ${this.name}`)
    }

    const firstInterpreted = this.firstExpression.interpret(
      model,
      variableAssignment
    )
    const tupleCandidates = relation[firstInterpreted]
    if (!tupleCandidates) {
      return false
    }

    const secondInterpreted = this.secondExpression.interpret(
      model,
      variableAssignment
    )
    return tupleCandidates.has(secondInterpreted)
  }

  public toFormattedString(): string {
    return `${
      this.name
    }(${this.firstExpression.toFormattedString()}, ${this.secondExpression.toFormattedString()})`
  }

  public toTree(): TreeNode {
    return {
      symbol: this.toFormattedString(),
      children: [],
    }
  }
}

export class EqualityRelation implements Relation {
  public constructor(
    public readonly firstExpression: Expression,
    public readonly secondExpression: Expression
  ) {}

  public evaluate(
    model: Model,
    variableAssignment: VariableAssignment
  ): boolean {
    const firstInterpreted = this.firstExpression.interpret(
      model,
      variableAssignment
    )
    const secondInterpreted = this.secondExpression.interpret(
      model,
      variableAssignment
    )
    return firstInterpreted === secondInterpreted
  }

  public toFormattedString(): string {
    return `${this.firstExpression.toFormattedString()} = ${this.firstExpression.toFormattedString()}`
  }

  public toTree(): TreeNode {
    return {
      symbol: this.toFormattedString(),
      children: [],
    }
  }
}

export interface Expression extends FOLFragment {
  interpret(model: Model, variableAssignment: VariableAssignment): number
}

export class Constant implements Expression {
  public constructor(public readonly name: string) {}

  public toFormattedString(): string {
    return this.name
  }

  public interpret(model: Model): number {
    const interpreted = model.functions.constants[this.name]
    if (interpreted === undefined) {
      throw new Error(`Missing constant ${this.name}`)
    }
    return interpreted
  }
}

export class UnaryFunction implements Expression {
  public constructor(
    public readonly name: string,
    public readonly inner: Expression
  ) {}

  public toFormattedString(): string {
    return `${this.name}(${this.inner.toFormattedString()})`
  }

  public interpret(
    model: Model,
    variableAssignment: VariableAssignment
  ): number {
    const unaryFunction = model.functions.unary[this.name]
    if (unaryFunction === undefined) {
      throw new Error(`Missing unary function ${this.name}`)
    }
    const innerInterpreted = this.inner.interpret(model, variableAssignment)
    const interpreted = unaryFunction[innerInterpreted]
    if (interpreted === undefined) {
      throw new Error(
        `Non-total unary function ${this.name}(${innerInterpreted})`
      )
    }
    return interpreted
  }
}

export class BinaryFunction implements Expression {
  public constructor(
    public readonly name: string,
    public readonly firstArgument: Expression,
    public readonly secondArgument: Expression
  ) {}

  public toFormattedString(): string {
    return `${
      this.name
    }(${this.firstArgument.toFormattedString()}, ${this.secondArgument.toFormattedString()})`
  }

  public interpret(
    model: Model,
    variableAssignment: VariableAssignment
  ): number {
    const binaryFunction = model.functions.binary[this.name]
    if (binaryFunction === undefined) {
      throw new Error(`Missing binary function ${this.name}`)
    }
    const firstInterpreted = this.firstArgument.interpret(
      model,
      variableAssignment
    )
    const curriedFunction = binaryFunction[firstInterpreted]
    if (curriedFunction === undefined) {
      throw new Error(
        `Non-total binary function ${this.name}(${firstInterpreted}, ...)`
      )
    }
    const secondInterpreted = this.secondArgument.interpret(
      model,
      variableAssignment
    )
    const interpreted = curriedFunction[secondInterpreted]
    if (interpreted === undefined) {
      throw new Error(
        `Non-total binary function ${this.name}(${firstInterpreted}, ${secondInterpreted})`
      )
    }
    return interpreted
  }
}

export class BoundVariable implements Expression {
  public constructor(public readonly name: string) {}

  public toFormattedString(): string {
    return this.name
  }

  public interpret(_: Model, variableAssignment: VariableAssignment): number {
    return variableAssignment[this.name]
  }
}
