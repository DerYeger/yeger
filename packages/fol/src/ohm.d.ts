import { Node } from 'ohm-js'
import { BoundVariable, Term, Formula } from '~/model'

export type BoundVariables = Set<String>

declare module 'ohm-js' {
  interface Node {
    args: {
      boundVariables: BoundVariables
    }
    parseFormula(boundVariables: BoundVariables): Formula
    parseString(): string
    parseTerm(boundVariables: BoundVariables): Term
    parseTermList(boundVariables: BoundVariables): Term[]
    parseVariable(boundVariables: BoundVariables): BoundVariable
  }

  interface Dict {
    parseFormula(boundVariables: BoundVariables): Formula
  }
}
