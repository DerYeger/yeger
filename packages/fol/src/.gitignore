fol.ohm-bundle.d.ts
fol.ohm-bundle.js
