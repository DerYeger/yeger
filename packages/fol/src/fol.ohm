FOL {
  Formula = BiImplFormula

  BiImplFormula
    = BiImplFormula "<->" ImplFormula -- biimpl
    | ImplFormula

  ImplFormula
    = ImplFormula "->" OrFormula -- impl
    | OrFormula

  OrFormula
    = OrFormula ("||" | "|") AndFormula -- or
    | AndFormula

  AndFormula
    = AndFormula ("&&" | "&") UnaryFormula -- and
    | UnaryFormula

  UnaryFormula
    = "(" Formula ")" -- paren
    | "forall" BoundVariable "." Formula -- universal
    | "exists" BoundVariable "." Formula -- existential
    | "!" UnaryFormula -- not
    | Relation
    | BooleanLiteral

  RelationName (a relation)
    = ident

  FunctionName (a function)
    = ident

  BoundVariable (a bound variable)
    = ident

  Relation
    = EqualityRelation
    | UnaryRelation
    | BinaryRelation

  BooleanLiteral
    = True
    | False

  UnaryRelation = RelationName "(" Expression ")"

  BinaryRelation = RelationName "(" Expression "," Expression ")"

  EqualityRelation = Expression "=" Expression

  True = "tt"

  False = "ff"

  Expression
    = UnaryFunction
    | BinaryFunction
    | ConstantOrBoundVariable

  ConstantOrBoundVariable (a bound variable or constant)
    = ident

  ident (an identifier)
    = letter+

  UnaryFunction = FunctionName "(" Expression ")"

  BinaryFunction = FunctionName "(" Expression "," Expression ")"
}
