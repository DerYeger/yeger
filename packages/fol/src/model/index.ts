export * from '~/model/formula'
export * from '~/model/model'
export * from '~/model/term'
