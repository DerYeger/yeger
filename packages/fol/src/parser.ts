import type { FOLSemantics } from '~/fol.ohm-bundle'
import grammar from '~/fol.ohm-bundle'
import type { Expression, Formula } from '~/model'
import {
  AndFormula,
  BiImplFormula,
  BinaryFunction,
  BinaryRelation,
  BooleanLiteral,
  BoundVariable,
  Constant,
  EqualityRelation,
  ExistentialQuantorFormula,
  ImplFormula,
  NotFormula,
  OrFormula,
  ParenthesizedFormula,
  UnaryFunction,
  UnaryRelation,
  UniversalQuantorFormula,
} from '~/model'

const semantics: FOLSemantics = grammar
  .createSemantics()
  .addOperation<string>('parseString()', {
    ident(_) {
      return this.sourceString
    },
  })
  .addOperation<BoundVariable>('parseVariable(boundVariables)', {
    BoundVariable(name) {
      return new BoundVariable(name.parseString())
    },
  })
  .addOperation<Expression>('parseExpression(boundVariables)', {
    ConstantOrBoundVariable(name) {
      const parsedName = name.parseString()
      if (this.args.boundVariables.has(parsedName)) {
        return new BoundVariable(parsedName)
      }
      return new Constant(parsedName)
    },
    UnaryFunction(name, _leftParen, inner, _rightParen) {
      return new UnaryFunction(
        name.parseString(),
        inner.parseExpression(this.args.boundVariables)
      )
    },
    BinaryFunction(
      name,
      _leftParen,
      firstArgument,
      _comma,
      secondArgument,
      _rightParen
    ) {
      return new BinaryFunction(
        name.parseString(),
        firstArgument.parseExpression(this.args.boundVariables),
        secondArgument.parseExpression(this.args.boundVariables)
      )
    },
  })
  .addOperation<Formula>('parseFormula(boundVariables)', {
    OrFormula_or(left, _operator, right) {
      return new OrFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    AndFormula_and(left, _operator, right) {
      return new AndFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    ImplFormula_impl(left, _operator, right) {
      return new ImplFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    BiImplFormula_biimpl(left, _operator, right) {
      return new BiImplFormula(
        left.parseFormula(this.args.boundVariables),
        right.parseFormula(this.args.boundVariables)
      )
    },
    UnaryFormula_paren(_leftParen, inner, _rightParen) {
      return new ParenthesizedFormula(
        inner.parseFormula(this.args.boundVariables)
      )
    },
    UnaryFormula_not(_not, inner) {
      return new NotFormula(inner.parseFormula(this.args.boundVariables))
    },
    UnaryFormula_universal(_quantor, variable, _dot, inner) {
      const parsedVariable = variable.parseVariable(this.args.boundVariables)
      const newBoundVariables = new Set([
        ...this.args.boundVariables,
        parsedVariable.name,
      ])
      return new UniversalQuantorFormula(
        parsedVariable,
        inner.parseFormula(newBoundVariables)
      )
    },
    UnaryFormula_existential(_quantor, variable, _dot, inner) {
      const parsedVariable = variable.parseVariable(this.args.boundVariables)
      const newBoundVariables = new Set([
        ...this.args.boundVariables,
        parsedVariable.name,
      ])
      return new ExistentialQuantorFormula(
        parsedVariable,
        inner.parseFormula(newBoundVariables)
      )
    },
    UnaryRelation(name, _leftParen, expression, _rightParen) {
      return new UnaryRelation(
        name.parseString(),
        expression.parseExpression(this.args.boundVariables)
      )
    },
    BinaryRelation(
      name,
      _leftParen,
      firstExpression,
      _comma,
      secondExpression,
      _rightParen
    ) {
      return new BinaryRelation(
        name.parseString(),
        firstExpression.parseExpression(this.args.boundVariables),
        secondExpression.parseExpression(this.args.boundVariables)
      )
    },
    EqualityRelation(firstExpression, _equality, secondExpression) {
      return new EqualityRelation(
        firstExpression.parseExpression(this.args.boundVariables),
        secondExpression.parseExpression(this.args.boundVariables)
      )
    },
    True(_) {
      return BooleanLiteral.True
    },
    False(_) {
      return BooleanLiteral.False
    },
  })

export function parse(formula: string): Formula {
  const matchResult = grammar.match(formula, 'Formula')
  return semantics(matchResult).parseFormula(new Set())
}
