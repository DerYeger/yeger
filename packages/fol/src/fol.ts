import type { Result } from 'resumon'

import grammar from '~/fol.ohm-bundle'
import type { Model } from '~/model'
import { parse } from '~/parser'

export * from '~/model'
export * from '~/validator'
export * from '~/fol.ohm-bundle'
export type { MatchResult } from 'ohm-js'

function evaluate(model: Model, formula: string): Result<boolean, string> {
  return parse(formula).map((parsedFormula) =>
    parsedFormula.evaluate(model, {})
  )
}

export const FOL = {
  evaluate,
  match: (formula: string) => grammar.match(formula),
  parse,
} as const
