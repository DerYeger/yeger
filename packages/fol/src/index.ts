import type { Result } from 'resumon'
import { Err, Ok } from 'resumon'

import grammar from '~/fol.ohm-bundle'
import type {
  Formula,
  Model,
  ModelCheckerMode,
  ModelCheckerTrace,
} from '~/model'
import { parse } from '~/parser'

export * from '~/fol.ohm-bundle'
export * from '~/model'
export * from '~/validator'
export type { MatchResult } from 'ohm-js'

function evaluate(model: Model, formula: string): Result<boolean, string> {
  return parse(formula).andThen<boolean>((parsedFormula) => {
    try {
      return new Ok(parsedFormula.evaluate(model, {}))
    } catch (err) {
      return new Err((err as Error).message)
    }
  })
}

function traceEvaluation(
  mode: ModelCheckerMode,
  expected: boolean,
  model: Model,
  formula: string,
): Result<ModelCheckerTrace, string> {
  return parse(formula).andThen<ModelCheckerTrace>((parsedFormula) => {
    try {
      return new Ok(parsedFormula.traceEvaluation(mode, expected, model, {}))
    } catch (err) {
      return new Err((err as Error).message)
    }
  })
}

function trace(
  mode: ModelCheckerMode,
  expected: boolean,
  model: Model,
  formula: Formula,
): Result<ModelCheckerTrace, string> {
  try {
    return new Ok(formula.traceEvaluation(mode, expected, model, {}))
  } catch (err) {
    return new Err((err as Error).message)
  }
}

export const FOL = {
  evaluate,
  match: (formula: string) => grammar.match(formula),
  parse,
  trace,
  traceEvaluation,
} as const
