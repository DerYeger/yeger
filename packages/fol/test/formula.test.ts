import { describe, expect, it } from 'vitest'

import { FOL } from '../src/index'
import { TestData } from './test-utils'

describe('FOL formula', () => {
  it.each(TestData.validFormulas)(
    'has correct formattedString for "%s"',
    (formula) => {
      const result = FOL.parse(formula)
      expect(result.get().toFormattedString()).toMatchSnapshot()
      // console.log(result)
    },
  )
})
