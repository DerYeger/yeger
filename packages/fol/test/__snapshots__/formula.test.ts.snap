// Vitest Snapshot v1

exports[`FOL formula > has correct formattedString for "!(tt && !(ff) && tt)" 1`] = `"¬((⊤ ∧ ¬((⊥)) ∧ ⊤))"`;

exports[`FOL formula > has correct formattedString for "!(tt && !ff) && tt" 1`] = `"¬((⊤ ∧ ¬(⊥))) ∧ ⊤"`;

exports[`FOL formula > has correct formattedString for "!A(x)" 1`] = `"¬(A(x))"`;

exports[`FOL formula > has correct formattedString for "!ff" 1`] = `"¬(⊥)"`;

exports[`FOL formula > has correct formattedString for "!tt && !(ff && tt)" 1`] = `"¬(⊤) ∧ ¬((⊥ ∧ ⊤))"`;

exports[`FOL formula > has correct formattedString for "!tt" 1`] = `"¬(⊤)"`;

exports[`FOL formula > has correct formattedString for "(forall x. A(x)) && (exists x. B(x))" 1`] = `"(∀x. A(x)) ∧ (∃x. B(x))"`;

exports[`FOL formula > has correct formattedString for "(tt || ff) && tt" 1`] = `"(⊤ ∨ ⊥) ∧ ⊤"`;

exports[`FOL formula > has correct formattedString for "A(x) & B(x)" 1`] = `"A(x) ∧ B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x) && B(x)" 1`] = `"A(x) ∧ B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x) <-> B(x)" 1`] = `"A(x) ↔ B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x) | B(x)" 1`] = `"A(x) ∨ B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x) || B(x)" 1`] = `"A(x) ∨ B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x) -> B(x)" 1`] = `"A(x) → B(x)"`;

exports[`FOL formula > has correct formattedString for "A(x)" 1`] = `"A(x)"`;

exports[`FOL formula > has correct formattedString for "A(x,y)" 1`] = `"A(x, y)"`;

exports[`FOL formula > has correct formattedString for "Relation(x)" 1`] = `"Relation(x)"`;

exports[`FOL formula > has correct formattedString for "a = b" 1`] = `"a = a"`;

exports[`FOL formula > has correct formattedString for "a = f(a,b)" 1`] = `"a = a"`;

exports[`FOL formula > has correct formattedString for "exists a . tt" 1`] = `"∃a. ⊤"`;

exports[`FOL formula > has correct formattedString for "exists x. A(x)" 1`] = `"∃x. A(x)"`;

exports[`FOL formula > has correct formattedString for "f(a) = a" 1`] = `"f(a) = f(a)"`;

exports[`FOL formula > has correct formattedString for "f(g(x)) = y" 1`] = `"f(g(x)) = f(g(x))"`;

exports[`FOL formula > has correct formattedString for "f(x) = y" 1`] = `"f(x) = f(x)"`;

exports[`FOL formula > has correct formattedString for "ff" 1`] = `"⊥"`;

exports[`FOL formula > has correct formattedString for "forall a. tt" 1`] = `"∀a. ⊤"`;

exports[`FOL formula > has correct formattedString for "forall x. A(x)" 1`] = `"∀x. A(x)"`;

exports[`FOL formula > has correct formattedString for "forall x. exists y. forall z. f(x) = y && R(y,z) && R(z,y) || (f(x) = x -> f(z) = a <-> ff)" 1`] = `"∀x. ∃y. ∀z. f(x) = f(x) ∧ R(y, z) ∧ R(z, y) ∨ (f(x) = f(x) → f(z) = f(z) ↔ ⊥)"`;

exports[`FOL formula > has correct formattedString for "forall x. forall y. A(x, y)" 1`] = `"∀x. ∀y. A(x, y)"`;

exports[`FOL formula > has correct formattedString for "forall x. forall y. h(x,y) = x" 1`] = `"∀x. ∀y. h(x, y) = h(x, y)"`;

exports[`FOL formula > has correct formattedString for "myConstant = myConstant" 1`] = `"myConstant = myConstant"`;

exports[`FOL formula > has correct formattedString for "myRelation(MyFunction(someConstant, AnotherConstant))" 1`] = `"myRelation(MyFunction(someConstant, AnotherConstant))"`;

exports[`FOL formula > has correct formattedString for "tt && tt" 1`] = `"⊤ ∧ ⊤"`;

exports[`FOL formula > has correct formattedString for "tt <-> tt" 1`] = `"⊤ ↔ ⊤"`;

exports[`FOL formula > has correct formattedString for "tt || ff && tt" 1`] = `"⊤ ∨ ⊥ ∧ ⊤"`;

exports[`FOL formula > has correct formattedString for "tt || tt" 1`] = `"⊤ ∨ ⊤"`;

exports[`FOL formula > has correct formattedString for "tt -> tt" 1`] = `"⊤ → ⊤"`;

exports[`FOL formula > has correct formattedString for "tt" 1`] = `"⊤"`;
